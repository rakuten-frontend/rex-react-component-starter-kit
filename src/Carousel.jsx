import React from 'react';
import { string, func, element, oneOfType, arrayOf } from 'prop-types';
import './Carousel.scss';

export default function Carousel({ children, onClick, text, className }) {
  return (
    <div className={className} onClick={onClick} role="presentation">
      <h1>{text}</h1>
      {children}
    </div>
  );
}

Carousel.defaultProps = {
  children: null,
  text: 'Welcome to React',
  className: 'rex-carousel',
  onClick: () => null,
};

Carousel.propTypes = {
  children: oneOfType([string, element, arrayOf(element)]),
  text: string,
  className: string,
  onClick: func,
};
