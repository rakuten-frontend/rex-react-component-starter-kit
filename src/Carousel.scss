.rex-carousel {
  h1 {
    text-decoration: underline;
  }
}