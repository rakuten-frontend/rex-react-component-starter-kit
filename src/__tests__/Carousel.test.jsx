import { shallow } from 'enzyme';
import React from 'react';
import Carousel from '../Carousel';

describe('<Carousel />', () => {
  const params = {
    text: 'Hello ReX React Component Starter Kit!!',
  };
  const makeSubject = (props, children) =>
    shallow(<Carousel {...props}>{children}</Carousel>);

  beforeEach(() => {
    jest.resetModules();
  });

  it('should match snapshot', () => {
    const component = makeSubject(params);
    expect(component.html()).toMatchSnapshot();
  });

  it('should render default-props', () => {
    const component = makeSubject();

    // Text
    expect(component.text()).toBe('Welcome to React');
    // className
    expect(component.hasClass('rex-carousel')).toBe(true);
    // click, no side-effect
    const clickedComponent = component.simulate('click');
    expect(clickedComponent).toBe(component);
  });

  it('should render passed props', () => {
    const component = makeSubject({
      text: 'Hello Jest Test!!',
      className: 'my-custom-class',
    });

    // Text
    expect(component.text()).toBe('Hello Jest Test!!');
    // className
    expect(component.hasClass('my-custom-class')).toBe(true);
  });

  it('simulate click event', () => {
    const clickHandler = jest.fn();
    const component = makeSubject({
      onClick: clickHandler,
    });

    component.find('div').simulate('click');
    expect(clickHandler).toHaveBeenCalled();
  });

  it('should render a child', () => {
    const component = makeSubject(params, <div>child</div>);
    expect(component.contains(<div>child</div>)).toBe(true);

    // Same as above:
    const wrapper = shallow(
      <Carousel>
        <p className="child">I am a child</p>
      </Carousel>
    );
    expect(
      wrapper.containsAllMatchingElements([
        <p className="child">I am a child</p>,
      ])
    ).toBe(true);
  });

  it('should render children', () => {
    const wrapper = shallow(
      <Carousel>
        <p className="child">I am a child</p>
        <span className="child">I am a child, too</span>
      </Carousel>
    );
    expect(
      wrapper.containsAllMatchingElements([
        <p className="child">I am a child</p>,
        <span className="child">I am a child, too</span>,
      ])
    ).toBe(true);
  });
});
