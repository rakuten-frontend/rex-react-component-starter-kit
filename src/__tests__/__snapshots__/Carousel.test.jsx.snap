// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> should match snapshot 1`] = `"<div class=\\"rex-carousel\\" role=\\"presentation\\"><h1>Hello ReX React Component Starter Kit!!</h1></div>"`;
